### PR TITLE
refactor: Combine material, measurement and hole handling in Core CKF `filter`

### PR DESCRIPTION
**Pull request by andiwand**
Combines the handling of material, measurements and holes in the Core CKF.

---

This pull request includes several changes to the `CombinatorialKalmanFilter` class in the `Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp` file to improve error handling, state management, and logging. Additionally, there is a minor change in the `MeasurementSelector` class to handle empty measurement candidates more gracefully.

### Error Handling and State Management:

* Added checks and logging for empty active branches to ensure the filter stops correctly when no branches are active. (`Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp`) [[1]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09R583-R590) [[2]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09R660-R673)
* Modified the filter to reset and stop properly when all branches are stopped or when certain conditions are met, improving the robustness of the filtering process. (`Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp`) [[1]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L592-R600) [[2]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L623-L642)

### Logging Improvements:

* Enhanced logging to provide more detailed information about the state transitions and decisions made during the filtering process. (`Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp`) [[1]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L716-R761) [[2]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L749-L925) [[3]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L950-R907)

### Code Simplification and Cleanup:

* Simplified the handling of material and measurement surfaces by refactoring conditional checks and removing redundant code. (`Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp`) [[1]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L716-R761) [[2]](diffhunk://#diff-55dd901875f5a87b6a9fee495de245296efbfd45aa9f9483036bcd299a277e09L749-L925)

### Minor Change in MeasurementSelector:

* Updated the `MeasurementSelector::select` method to return a success result when no measurement candidates are found, instead of returning an error. (`Core/include/Acts/TrackFinding/MeasurementSelector.ipp`)

### CODE REVIEW

**Comment by andiwand** [comment]
cc @paulgessinger @goetzgaycken

**Comment by andiwand** [comment]
I observe similar CPU performance against main

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 1]
```diff
@@ -713,217 +716,165 @@ class CombinatorialKalmanFilter {
       using PM = TrackStatePropMask;
 
       bool isSensitive = surface->associatedDetectorElement() != nullptr;
-      bool isMaterial = surface->surfaceMaterial() != nullptr;
-
-      std::size_t nBranchesOnSurface = 0;
+      bool hasMaterial = surface->surfaceMaterial() != nullptr;
+      bool isMaterial = hasMaterial && !isSensitive;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:720]: `isMaterialOnly`?

Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:720]: `isMaterialOnly`?

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 2]
```diff
@@ -713,217 +716,165 @@ class CombinatorialKalmanFilter {
...
+        ACTS_VERBOSE("Material surface " << surface->geometryId()
+                                         << " detected.");
+      } else {
+        ACTS_VERBOSE("Passive surface " << surface->geometryId()
+                                        << " detected.");
+        return Result<void>::success();
+      }
 
-        // Transport the covariance to the surface
+      using SourceLinkRange = decltype(m_sourceLinkAccessor(*surface));
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:735]: We don't want to write this out at this point?

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 3]
```diff
@@ -713,217 +716,165 @@ class CombinatorialKalmanFilter {
...
-        TrackStatesResult tsRes = trackStateCandidateCreator(
+      // Create trackstates for all source links (will be filtered later)
+      using TrackStatesResult =
+          Acts::Result<CkfTypes::BranchVector<TrackIndexType>>;
+      TrackStatesResult tsRes = TrackStatesResult::success({});
+      if (hasMeasurements) {
+        tsRes = trackStateCandidateCreator(
             state.geoContext, *calibrationContextPtr, *surface, boundState,
-            slBegin, slEnd, prevTip, *result.trackStates,
+            slRange->first, slRange->second, prevTip, *result.trackStates,
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:779]: I'd probably still use a structured binding here.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 4]
```diff
@@ -713,217 +716,165 @@ class CombinatorialKalmanFilter {
...
-        if (!boundStateRes.ok()) {
-          return boundStateRes.error();
-        }
-        auto& boundState = *boundStateRes;
-        auto& [boundParams, jacobian, pathLength] = boundState;
-        boundParams.covariance() = state.stepping.cov;
-
-        // Add a hole or material track state to the multitrajectory
+        // Add a hole track state to the multitrajectory
         TrackIndexType currentTip = addNonSourcelinkState(
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:820]: I don't see any other call to `addNonSourcelinkState` now. Are material-only surfaces not recorded as track states anymore? Don't we also get here if `expectMeasurements == false`, which case it's a material state? If so I suppose the command above is incorrect?

File: Core/include/Acts/TrackFinding/MeasurementSelector.ipp [hunk 5]
```diff
@@ -21,9 +21,10 @@ MeasurementSelector::select(
 
   ACTS_VERBOSE("Invoked MeasurementSelector");
 
-  // Return error if no measurement
+  // Return if no measurement
   if (candidates.empty()) {
-    return CombinatorialKalmanFilterError::MeasurementSelectionFailed;
+    return Result::success(
+        std::make_pair(candidates.begin(), candidates.end()));
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/MeasurementSelector.ipp:27]:
```suggestion
        std::pair{candidates.begin(), candidates.end())};
```

if possible.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp (hunk 2 above)
Comment (andiwand) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:735]: You mean explicitly writing this type out? I think we have no access to the range type or iterator types but maybe I am wrong

Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:735]: Ah right, I think I got confused.

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [e3b33ccf43a02d41e59dd9e9b666f28b28d61b48]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684677)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684676)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684675)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684674)||
|:yellow_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684673)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684672)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684671)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684670)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684669)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684668)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684667)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684666)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684665)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684664)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684663)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684662)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684661)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684660)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684659)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684658)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684657)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684656)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684655)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684654)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684653)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684652)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684651)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684650)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684649)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684648)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684647)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684646)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44684645)||

**Pull request merged**